Makefile

```diff
@@ -0,0 +1,104 @@
+# Copyright 2018 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+TARGET = federation-v2
+GOTARGET = github.com/kubernetes-sigs/$(TARGET)
+REGISTRY ?= quay.io/kubernetes-multicluster
+IMAGE = $(REGISTRY)/$(TARGET)
+DIR := ${CURDIR}
+DOCKER ?= docker
+
+GIT_VERSION ?= $(shell git describe --always --dirty)
+GIT_TAG ?= $(shell git describe --tags --exact-match 2>/dev/null)
+GIT_HASH ?= $(shell git rev-parse HEAD)
+BUILDDATE = $(shell date -u +'%Y-%m-%dT%H:%M:%SZ')
+
+GIT_TREESTATE = "clean"
+DIFF = $(shell git diff --quiet >/dev/null 2>&1; if [ $$? -eq 1 ]; then echo "1"; fi)
+ifeq ($(DIFF), 1)
+    GIT_TREESTATE = "dirty"
+endif
+
+ifneq ($(VERBOSE),)
+VERBOSE_FLAG = -v
+endif
+BUILDMNT = /go/src/$(GOTARGET)
+BUILD_IMAGE ?= golang:1.10.3
+
+CONTROLLER_TARGET = bin/controller-manager
+KUBEFED2_TARGET = bin/kubefed2
+
+LDFLAG_OPTIONS = -ldflags "-X github.com/kubernetes-sigs/federation-v2/pkg/version.version=$(GIT_VERSION) \
+                      -X github.com/kubernetes-sigs/federation-v2/pkg/version.gitCommit=$(GIT_HASH) \
+                      -X github.com/kubernetes-sigs/federation-v2/pkg/version.gitTreeState=$(GIT_TREESTATE) \
+                      -X github.com/kubernetes-sigs/federation-v2/pkg/version.buildDate=$(BUILDDATE)"
+BUILDCMD_CONTROLLER = go build -o $(CONTROLLER_TARGET) $(VERBOSE_FLAG) $(LDFLAG_OPTIONS)
+BUILDCMD_KUBEFED2 = go build -o $(KUBEFED2_TARGET) $(VERBOSE_FLAG) $(LDFLAG_OPTIONS)
+
+BUILD_CONTROLLER = $(BUILDCMD_CONTROLLER) cmd/controller-manager/main.go
+BUILD_KUBEFED2 = $(BUILDCMD_KUBEFED2) cmd/kubefed2/kubefed2.go
+
+TESTARGS ?= $(VERBOSE_FLAG) -timeout 60s
+TEST_PKGS ?= $(GOTARGET)/cmd/... $(GOTARGET)/pkg/...
+TEST_CMD = go test $(TESTARGS)
+TEST = $(TEST_CMD) $(TEST_PKGS) 
+
+VET = go vet $(TEST_PKGS)
+FMT = go fmt $(TEST_PKGS)
+
+DOCKER_BUILD ?= $(DOCKER) run --rm -v $(DIR):$(BUILDMNT) -w $(BUILDMNT) $(BUILD_IMAGE) /bin/sh -c
+
+# TODO (irfanurrehman): can add local compile, and auto-generate targets also if needed
+.PHONY: all container push clean controller kubefed2 test local-test vet fmt
+
+all: container controller kubefed2
+
+local-test:
+	$(TEST)
+
+# Unit tests 
+test: controller kubefed2 vet
+	$(DOCKER_BUILD) '$(TEST)'
+
+vet:
+	$(DOCKER_BUILD) '$(VET)'
+
+fmt:
+	$(FMT)
+
+container:
+	$(DOCKER) build -f Dockerfile.controller \
```

Comment:
The image build in travis is using `images/federation-v2/Dockerfile` to do the build. We should pick one - the dockerfile referenced here, or the dockerfile the image build script is already using.

Reply:
For reference, the `images/federation-v2/Dockerfile` doesn't do any of the go tests or builds done in Dockerfile.controller because they're already done earlier in the travis build with `precommit.sh`. Since it's possible to copy the binaries created with `download-binaries.sh` into the path for the docker build, there's no need to run through Dockerfile.controller again.

Reply:
I found both Dockerfiles, while updating the Makefile; Both have flaws. Tests are not needed as part of `Dockerfile.controller` and `images/federation-v2/Dockerfile` can only be invoked from `scripts/imagebuild.sh`, because the script basically copies/provides the `controller-manager` binary to be copied into the image. I think `Dockerfile.controller` is better as it builds the binary and can be invoked without dependency. I also think that this change could be done subsequently, as followup, where we can rather use `make container` and `make push` to build and push the image from CI and should not be a blocker for making this release. When both merge, tests could also be removed from `Dockerfile.controller`.
Meanwhile, the current use of `Dockerfile.controller` is to provide the user a mechanism to build their own image.

Reply:
I would really prefer to avoid having two Dockerfiles - I don't have a strong opinion on which one we use, but I don't want to have two different Dockerfiles and ways of building the image. Let's pick one now and change whatever needs to in this PR.

How does that sound?

Reply:
Also, I don't expect users to normally want to build their own image. If you're building your own image locally, you're a federation developer, IMO